### ISSUE TITLE
JSpecify: false positive for lambda returning null casting to function returning null

### ISSUE THREAD

**Issue by unv-unv** [uber/NullAway#1802]
```java
@NullMarked
public class UnmodifiableHashMapTest {
    public static <E> Iterator<E> iterateWhileNonNull(@Nullable E first, Function<? super E, @Nullable ? extends E> generator) {
        if (first == null) {
            return List.<E>of().iterator();
        }
        return new Iterator<>() {
            E current = first;

            @Override
            public boolean hasNext() {
                E next = generator.apply(current);
                return next != null;
            }

            @Override
            public E next() {
                E next = generator.apply(current);
                if (next == null) {
                    throw new NoSuchElementException();
                }
                current = next;
                return next;
            }
        };
    }

    Iterator<Object> test() {
        return iterateWhileNonNull(new Object(), o -> null /* BUG: returning @Nullable expression from method with @NonNull return type */);
    }
```

Shows compilation error:
```
[ERROR] Test.java:[29,49] [NullAway] returning @Nullable expression from method with @NonNull return type
```

NullAway 0.14.1, Error Prone 2.50.0.

**Comment by jeffrey-easyesi** [comment]
`@Nullable ? extends E` should be `? extends @Nullable E`; there's a specific error for this with NullAway:HandleWildcardGenerics enabled
```
[NullAway] illegal location for annotation: nullness annotations cannot be written directly on a wildcard; annotate the wildcard bound instead
    public static <E> Iterator<E> iterateWhileNonNull(@Nullable E first, Function<? super E, @Nullable ? extends E> generator) {
                                                                                             ^
```

**Comment by msridhar** [comment]
@unv-unv are you trying to enroll a codebase onto NullAway from scratch?  If so, I recommend trying the `JSpecifyExperimental` flag (which also turns on `HandleWildcardGenerics`).  It'll handle cases like these better.

**Comment by unv-unv** [comment]
> [@unv-unv](https://github.com/unv-unv) are you trying to enroll a codebase onto NullAway from scratch? If so, I recommend trying the `JSpecifyExperimental` flag (which also turns on `HandleWildcardGenerics`). It'll handle cases like these better.

I'm already doing this for several months. My target is to convert all the codebase of Consulo IDE (IntelliJ IDEA fork, 20+ thousands of classes) to NullAway module by module. I already converted 60 of 200+ modules. I have just upgraded to NullAway 0.14.1. Half of the NullAway suppressions are gone thanks to your work. I'll try a new flag, thanks.

**Comment by msridhar** [comment]
Fair warning, the new flag will cause a bunch of new warnings in the modules you already migrated, due to incorporating nullability models for JDK classes / methods.  But, I think it's worth it.

@unv-unv can we close this issue, or is there still some concern here?

**Comment by msridhar** [comment]
I'm going to close assuming this is addressed, can re-open if needed
